### DISCUSSION TITLE
How to update atom value in window.postMessage event listener

### DISCUSSION THREAD

**Discussion by akaSybe** [pmndrs/jotai#1117] [Q&A] [answered]
Hey! Thank you very much for such cool library!

I have an atom and event listener for `message` type (imagine that other tab sends some data using `window.postMessage`) and I want to update my atom value on every event handler call. Can't figure it out how to make it work.

```
import { atom } from "jotai";

const store = atom(0);

window.addEventListener("message", (msg) => {
  const data = message.data; // 1
  // want to update an atom value from 0 to 1
});

window.postMessage(1);
```

P.S. The requirement is to update the value outside of React component

**Comment by dai-shi** [comment, marked as answer]
One easy solution is to use `useEffect`.

```js
import { atom, useSetAtom } from "jotai";

const store = atom(0);

const useSyncStore = () => {
  const set = useSetAtom(store);
  useEffect(() => {
    const callback = (msg) => {
      const data = message.data;
      set(data);
    });
    window.addEventListener("message", callback);
    callback(...); // fire once if needed
    return () => window.removeEventListener("message", calback);
  }, [set]);
};
```

and use `useSyncStore` somewhere appropriate.

----
Alternatively, we have `atom.onMount` feature, which is meant to be mainly used by utils.

```js
const store = atom(0);
store.onMount = (set) => {
  const callback = (msg) => {
    const data = message.data;
    set(data);
  });
  window.addEventListener("message", callback);
  callback(...); // fire once if needed
  return () => window.removeEventListener("message", calback);
};
```

**Reply by Jaaneek** [reply]
Hi, Is there a way to send big objects this way? Whenever we use postMessage on web workers/shared workers. postMessage copies the objects, which results in the receiving atom updating its whole content and triggering rerenders. Can we somehow send only what was updated and then update it on the other end? 

Without writing any reducers, just do it automatically, maybe some new API? 

Something like:

(updateEvent) => atom.consumeUpdateEvent(updateEvent) 


Do you have any guidance on how to synchronize the same exact atom in component and in Iframe that is inside that component?

**Reply by AjaxSolutions** [reply]
"Do you have any guidance on how to synchronize the same exact atom in component and in Iframe that is inside that component?"

Can you use atomWithBroadcast?

https://jotai.org/docs/recipes/atom-with-broadcast

**Reply by Jaaneek** [reply]
@AjaxSolutions Atom with broadcast is just sending a whole object with new reference every time. I don't want the whole object to be replaced. I only want the parts that were replaced. Right now I achieved that by using a lot of immer with createPatches but would love to have it as a mechanism in jotai instead that would be more performant
